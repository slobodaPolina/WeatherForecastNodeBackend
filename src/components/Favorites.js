import React from 'react';
import Forecast from './Forecast';
import SmallPreview from './SmallPreview';
import Loading from './Loading';
import { getCityByName, getTheFavorites, addCityToTheFavorites, deleteCityFromTheFavorites } from "../cityAdder";

export class Favorites extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            city: "", // it is the text from the searchBar
            isCityValid: true, // is the city user tried t add valid
            favorites: [] // array of objects of favorite cities. We keep only names in db
        };
        this.handleSearch = this.handleSearch.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
        getTheFavorites(
            (data) => {
                data.forEach(name => {
                    getCityByName(
                        name,
                        (data) => {
                            this.addToTheFavorites(Object.assign(data, { loaded: true }));
                        },
                        () => {
                            console.error("Error getting favorite city " + name + " from weather API");
                            this.addToTheFavorites({
                                name,
                                loaded: false
                            });
                        }
                    );
                });
            },
            () => console.error("Problems getting list of favorite cities from db!")
        );
    }

    render() {
        let msg = this.state.isCityValid ? "" : <div className="error-message">Seems like the city is not valid</div>;
        return (
            <div>
                <div>
                    <div className="search-bar row">
                        <div className="bigText">Избранное</div>
                        <div>
                            <form onSubmit={this.handleSearch}>
                                <input id="search-input" type="search"
                                       placeholder="Добавить новый город" value={this.state.city} onChange={this.handleInputChange}/>
                                <button className="search-button roundButton">+</button>
                            </form>
                        </div>
                    </div>
                    {msg}
                </div>
                <div className="favorites">
                    {this.state.favorites.map((city) => {
                        return (
                            <div className="smallCityCard" key={city.name}>
                                <SmallPreview city={city} removeCity={this.removeGenerator(city.name)}/>
                                {city.loaded ? <Forecast data={city}/> : <Loading />}
                            </div>
                        )
                    })}
                </div>
            </div>
        );
    }

    handleInputChange(event) {
        this.setState({city: event.target.value});
    }

    handleSearch(event) {
        const getCallback = (isSuccess) => () => {
            this.setState({isCityValid: isSuccess});
            // todo remove loading card anyway
            if (isSuccess) { // means we got data from weather service
                addCityToTheFavorites(
                    this.state.city,
                    () => {
                        this.setState({favorites: this.state.favorites.push(this.state.city)});
                        this.setState({city: ""});
                    },
                    () => {
                        console.error('error during saving favorite city in db!');
                        this.setState({isCityValid: false});
                    }
                );
            }
        };

        event.preventDefault();
        if (this.props.favorites.find(city => city.name === this.state.city)) { // if we already have it we wont add it again
            this.setState({isCityValid: false});
        } else {
            // todo start the loader on the citycard
            // load real data and display it
            getCityByName(
                this.state.city,
                getCallback(true),
                getCallback(false)
            );
        }
    }

    removeGenerator(cityName) {
        return () => deleteCityFromTheFavorites(
            cityName,
            () => {
                this.setState({
                    favorites: this.state.favorites.filter(city => city.name !== cityName)
                })
            },
            () => { console.error("Problems removing city " + cityName + " from db!") }
         );
    }

    addToTheFavorites(obj) {
        this.setState({
            favorites: this.state.favorites.push(obj)
        });
    }
}

export default Favorites;
