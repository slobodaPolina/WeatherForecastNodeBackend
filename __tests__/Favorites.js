import React from 'react';
import renderer from 'react-test-renderer';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import rootReducer from '../src/reducers';
import Favorites from '../src/components/Favorites';

const unloadedStore = createStore(
    rootReducer,
    {
        favorites: [
            {
                name: 'City 1',
                loaded: false,
            },
            {
                name: 'City 2',
                loaded: false,
            }
        ],
    }
);

const emptyStore = createStore(
    rootReducer,
    {favorites: []}
);

const loadedStore = createStore(
    rootReducer,
    {
        favorites: [
            {
                clouds: {
                    all: 40
                },
                cod: 200,
                coord: {
                    lat: 51.51,
                    lon: -0.13
                },
                dt: 1573292996,
                id: 2643743,
                loaded: true,
                main: {
                    humidity: 93,
                    pressure: 1008,
                    pressureAtm: "1.01",
                    pressureMmHg: "756.00",
                    temp: "2.48",
                    temp_max: 277.59,
                    temp_min: 274.26
                },
                name: "London",
                rain: {},
                sys: {
                    country: "GB",
                    id: 1414,
                    sunrise: 1573283248,
                    sunset: 1573316474,
                    type: 1
                },
                timezone: 0,
                visibility: 5000,
                weather: [{
                    description: "haze",
                    icon: "50d",
                    id: 721,
                    main: "Haze"
                }],
                wind: {
                    speed: 1.5
                }
            },
            {
                clouds: {
                    all: 40
                },
                cod: 200,
                coord: {
                    lat: 51.51,
                    lon: -0.13
                },
                dt: 1573292996,
                id: 2643743,
                loaded: true,
                main: {
                    humidity: 93,
                    pressure: 1008,
                    pressureAtm: "1.01",
                    pressureMmHg: "756.00",
                    temp: "2.48",
                    temp_max: 277.59,
                    temp_min: 274.26
                },
                name: "Paris",
                rain: {},
                weather: [{
                    description: "haze",
                    icon: "50d",
                    id: 721,
                    main: "Haze"
                }],
                wind: {
                    speed: 1.5
                }
            }
        ],
    }
);


it('renders unloaded favorites correctly', () => {
  const tree = renderer.create(
        <Provider store={unloadedStore}>
            <Favorites/>
        </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});


it('renders empty favorites correctly', () => {
  const tree = renderer.create(
        <Provider store={emptyStore}>
            <Favorites/>
        </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders loaded favorites correctly', () => {
  const tree = renderer.create(
        <Provider store={loadedStore}>
            <Favorites/>
        </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
