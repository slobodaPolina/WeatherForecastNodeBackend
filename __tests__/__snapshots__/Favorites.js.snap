// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty favorites correctly 1`] = `
<div
  className="favorites"
/>
`;

exports[`renders loaded favorites correctly 1`] = `
<div
  className="favorites"
>
  <div
    className="smallCityCard"
  >
    <div
      className="row SmallPreviewContainer"
    >
      <div
        className="width45 row"
      >
        <div
          className="cityTitle"
        >
          London
        </div>
        <div
          className="cityTemp"
        >
          2.48
           °C
        </div>
        <div
          className="width36px"
        >
          <img
            alt="Weather"
            className="weatherImage"
            src="/images/Haze.jpg"
          />
        </div>
      </div>
      <button
        className="roundButton removeButton"
        onClick={[Function]}
      >
        +
      </button>
    </div>
    <div
      className="forecastContainer column"
    >
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Ветер
        </div>
        <div
          className="paramValue"
        >
          1.5
           meter/sec
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Облачность
        </div>
        <div
          className="paramValue"
        >
          haze
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Давление
        </div>
        <div
          className="paramValue"
        >
          756.00
           mm Hg
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Влажность
        </div>
        <div
          className="paramValue"
        >
          93
          %
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Координаты
        </div>
        <div
          className="paramValue"
        >
          [
          -0.13
          , 
          51.51
          ]
        </div>
      </div>
    </div>
  </div>
  <div
    className="smallCityCard"
  >
    <div
      className="row SmallPreviewContainer"
    >
      <div
        className="width45 row"
      >
        <div
          className="cityTitle"
        >
          Paris
        </div>
        <div
          className="cityTemp"
        >
          2.48
           °C
        </div>
        <div
          className="width36px"
        >
          <img
            alt="Weather"
            className="weatherImage"
            src="/images/Haze.jpg"
          />
        </div>
      </div>
      <button
        className="roundButton removeButton"
        onClick={[Function]}
      >
        +
      </button>
    </div>
    <div
      className="forecastContainer column"
    >
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Ветер
        </div>
        <div
          className="paramValue"
        >
          1.5
           meter/sec
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Облачность
        </div>
        <div
          className="paramValue"
        >
          haze
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Давление
        </div>
        <div
          className="paramValue"
        >
          756.00
           mm Hg
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Влажность
        </div>
        <div
          className="paramValue"
        >
          93
          %
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Координаты
        </div>
        <div
          className="paramValue"
        >
          [
          -0.13
          , 
          51.51
          ]
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders unloaded favorites correctly 1`] = `
<div
  className="favorites"
>
  <div
    className="smallCityCard"
  >
    <div
      className="row SmallPreviewContainer"
    >
      <div
        className="width45 row"
      >
        <div
          className="cityTitle"
        >
          City 1
        </div>
        
        
      </div>
      <button
        className="roundButton removeButton"
        onClick={[Function]}
      >
        +
      </button>
    </div>
    <div>
      <div
        className="bigText"
      >
        Подождите, данные загружаются
      </div>
      <div
        className="refreshBanner"
      >
        <img
          alt="loading"
          className="refreshImage"
          src="/images/refresh.svg"
        />
      </div>
    </div>
  </div>
  <div
    className="smallCityCard"
  >
    <div
      className="row SmallPreviewContainer"
    >
      <div
        className="width45 row"
      >
        <div
          className="cityTitle"
        >
          City 2
        </div>
        
        
      </div>
      <button
        className="roundButton removeButton"
        onClick={[Function]}
      >
        +
      </button>
    </div>
    <div>
      <div
        className="bigText"
      >
        Подождите, данные загружаются
      </div>
      <div
        className="refreshBanner"
      >
        <img
          alt="loading"
          className="refreshImage"
          src="/images/refresh.svg"
        />
      </div>
    </div>
  </div>
</div>
`;
